File: .claude/skills/flow-engineer-skill/scripts/package_skill.py
#!/usr/bin/env python3
"""
Skill Packager - Creates a distributable .skill file

Usage:
    python package_skill.py <path/to/skill-folder> [output-directory]

Example:
    python package_skill.py .cursor/skills/code-review
    python package_skill.py .cursor/skills/code-review ./dist
"""

import sys
import zipfile
from pathlib import Path
from validate_skill import validate_skill


def package_skill(skill_path, output_dir=None):
    """
    Package a skill folder into a .skill file.

    Args:
        skill_path: Path to the skill folder
        output_dir: Optional output directory (defaults to cwd)

    Returns:
        Path to created .skill file, or None if error
    """
    skill_path = Path(skill_path).resolve()

    if not skill_path.exists():
        print(f"Error: Skill folder not found: {skill_path}")
        return None

    if not skill_path.is_dir():
        print(f"Error: Path is not a directory: {skill_path}")
        return None

    if not (skill_path / "SKILL.md").exists():
        print(f"Error: SKILL.md not found in {skill_path}")
        return None

    # Validate before packaging
    print("Validating skill...")
    valid, message = validate_skill(skill_path)
    if not valid:
        print(f"Validation failed: {message}")
        print("   Fix validation errors before packaging.")
        return None
    print(f"{message}\n")

    # Determine output location
    skill_name = skill_path.name
    if output_dir:
        output_path = Path(output_dir).resolve()
        output_path.mkdir(parents=True, exist_ok=True)
    else:
        output_path = Path.cwd()

    skill_filename = output_path / f"{skill_name}.skill"

    # Create .skill file (zip format)
    try:
        with zipfile.ZipFile(skill_filename, "w", zipfile.ZIP_DEFLATED) as zipf:
            for file_path in skill_path.rglob("*"):
                if file_path.is_file():
                    arcname = file_path.relative_to(skill_path.parent)
                    zipf.write(file_path, arcname)
                    print(f"  Added: {arcname}")

        print(f"\nPackaged skill to: {skill_filename}")
        return skill_filename

    except Exception as e:
        print(f"Error creating .skill file: {e}")
        return None


def main():
    if len(sys.argv) < 2:
        print(
            "Usage: python package_skill.py <path/to/skill-folder> [output-directory]"
        )
        print("\nExample:")
        print("  python package_skill.py .cursor/skills/code-review")
        print("  python package_skill.py .cursor/skills/code-review ./dist")
        sys.exit(1)

    skill_path = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) > 2 else None

    print(f"Packaging skill: {skill_path}")
    if output_dir:
        print(f"   Output directory: {output_dir}")
    print()

    result = package_skill(skill_path, output_dir)
    sys.exit(0 if result else 1)


if __name__ == "__main__":
    main()

File: .claude/skills/flow-engineer-skill/scripts/package_skill.ts
#!/usr/bin/env -S deno run -A
/**
 * Skill Packager - Creates a distributable .skill file
 *
 * Usage:
 *     deno run -A package_skill.ts <path/to/skill-folder> [output-directory]
 *
 * Example:
 *     deno run -A package_skill.ts .cursor/skills/code-review
 *     deno run -A package_skill.ts .cursor/skills/code-review ./dist
 */

import { basename, join, resolve } from "jsr:@std/path";
import { validateSkill } from "./validate_skill.ts";

export async function packageSkill(
  skillPath: string,
  outputDir?: string | null,
): Promise<string | null> {
  const resolvedSkillPath = resolve(skillPath);

  // Check skill folder exists
  try {
    const stat = await Deno.stat(resolvedSkillPath);
    if (!stat.isDirectory) {
      console.log(`Error: Path is not a directory: ${resolvedSkillPath}`);
      return null;
    }
  } catch {
    console.log(`Error: Skill folder not found: ${resolvedSkillPath}`);
    return null;
  }

  // Check SKILL.md exists
  try {
    await Deno.stat(join(resolvedSkillPath, "SKILL.md"));
  } catch {
    console.log(`Error: SKILL.md not found in ${resolvedSkillPath}`);
    return null;
  }

  // Validate before packaging
  console.log("Validating skill...");
  const [valid, message] = validateSkill(resolvedSkillPath);
  if (!valid) {
    console.log(`Validation failed: ${message}`);
    console.log("   Fix validation errors before packaging.");
    return null;
  }
  console.log(`${message}\n`);

  // Determine output location
  const skillName = basename(resolvedSkillPath);
  let outputPath: string;
  if (outputDir) {
    outputPath = resolve(outputDir);
    await Deno.mkdir(outputPath, { recursive: true });
  } else {
    outputPath = Deno.cwd();
  }

  const skillFilename = join(outputPath, `${skillName}.skill`);

  // Create .skill file using zip -r
  // The Python version archives with paths relative to skill_path.parent,
  // so the zip contains <skill-name>/... as the top-level directory.
  try {
    const parentDir = resolve(resolvedSkillPath, "..");
    const dirName = basename(resolvedSkillPath);

    const cmd = new Deno.Command("zip", {
      args: ["-r", skillFilename, dirName],
      cwd: parentDir,
      stdout: "piped",
      stderr: "piped",
    });

    const output = await cmd.output();

    if (!output.success) {
      const stderr = new TextDecoder().decode(output.stderr);
      console.log(`Error creating .skill file: zip command failed: ${stderr}`);
      return null;
    }

    // Print added files to match Python behavior
    const stdout = new TextDecoder().decode(output.stdout);
    // zip -r output lines like "  adding: dir/file (deflated 42%)"
    // Python prints "  Added: skill-name/file"
    for (const line of stdout.split("\n")) {
      const addMatch = line.match(/^\s*adding:\s+(.+?)\s+\(/);
      if (addMatch) {
        const arcname = addMatch[1];
        // Only print files, not directories (directories end with /)
        if (!arcname.endsWith("/")) {
          console.log(`  Added: ${arcname}`);
        }
      }
    }

    console.log(`\nPackaged skill to: ${skillFilename}`);
    return skillFilename;
  } catch (e) {
    console.log(`Error creating .skill file: ${e}`);
    return null;
  }
}

async function main(): Promise<void> {
  if (Deno.args.length < 1) {
    console.log(
      "Usage: python package_skill.py <path/to/skill-folder> [output-directory]",
    );
    console.log("\nExample:");
    console.log("  python package_skill.py .cursor/skills/code-review");
    console.log("  python package_skill.py .cursor/skills/code-review ./dist");
    Deno.exit(1);
  }

  const skillPath = Deno.args[0];
  const outputDir = Deno.args.length > 1 ? Deno.args[1] : null;

  console.log(`Packaging skill: ${skillPath}`);
  if (outputDir) {
    console.log(`   Output directory: ${outputDir}`);
  }
  console.log();

  const result = await packageSkill(skillPath, outputDir);
  Deno.exit(result ? 0 : 1);
}

if (import.meta.main) {
  main();
}
